### DISCUSSION TITLE
negative flow rate of inlet boundary

### DISCUSSION THREAD

**Discussion by LilHang** [PMEAL/OpenPNM#2817] [Q&A]
Hello community,

I used openPNM to predict the permeability of a fibrous material, pore network was extracted using `snow2` and 3 boundary voxels were added by default. I want to test the results of different conduit shape and geometric properties, the inlet flow rate was used to calculate permeability, but some of the results i got were negative, which suggests the boundary flow rate is negative. 

For example, the result is `1.59e-11 m^2` which is close to experiments if i choose `cones_and_cylinders` as conduit shape,  `pore.equivalent_diameter`, `throat.equivalent_diameter`, `throat.direct_length` as `pore.diameter`, `throat.diameter`, `throat.spacing` respectively. However, the result is `-9.85e-13 m^2` if `pore.inscribed_diameter` is choosen as `pore.diameter`(other remains the same) .  According to the doc, negative flow rate means flow into the pores, so i think inlet flow rate must be positive which suggests flow is leaving the pore, i can not figure out what happend here, is the negative inlet flow rate normal here ? 

some of my code is listed below.

```python
# network extracted, 3 boundary voxels added by default
snow_output = ps.networks.snow2(im, voxel_size=5e-6, parallelization=None)
pn = op.io.network_from_porespy(snow_output.network)

# size factors
pn["pore.diameter"] = pn["pore.inscribed_diameter"]
pn["throat.diameter"] = pn["throat.equivalent_diameter"]
pn["throat.spacing"] = pn["throat.direct_length"]
pn.add_model(propname="throat.hydraulic_size_factors",
             model=op.models.geometry.hydraulic_size_factors.cones_and_cylinders)
# condunctance
water.add_model(propname="throat.hydraulic_conductance",
                model=op.models.physics.hydraulic_conductance.generic_hydraulic)

# boundary conditions
inlet = pn.pores("ymin")  
outlet = pn.pores("ymax")
sf.set_value_BC(pores=inlet, values=1.)
sf.set_value_BC(pores=outlet, values=0.)

# caluclate inlet flow rate after run
Q = sf.rate(pores=inlet)[0]
```

**Comment by jgostick** [comment]
I'm guess that your pore and throat sizes were wonky, and overlapped a lot, so you ended up with negative conductance values.  Maybe to `plt.hist(phase['throat.hydraulic_conductance'], bins=50)` and see where they lie?

**Reply by LilHang** [reply]
Thanks for your quick reply professor. i follow your instructions and get two conductance histogram with different geometric properties. There are 60 and 373 negative conductance values, the second one match well with experiments. I look into source code of `size factors` and `generic_hydraulic` calculation, it seems that variables in the formula are all positive, i have no ideas how the negative sign comes in. should i have to fix this or just leave as it is and choose the shape and geometric properties that match the experiments well ？ 
<img width="589" alt="截屏2023-09-12 13 17 35" src="https://github.com/PMEAL/OpenPNM/assets/42541380/bce2664f-48d1-49d2-91c1-233ab8cbe437">
<img width="578" alt="截屏2023-09-12 13 27 24" src="https://github.com/PMEAL/OpenPNM/assets/42541380/8430a109-3c14-4fe4-bfce-a97b0ed39da0">

**Reply by gxw-ll** [reply]
Hi，prof, the snow2 algorithm is fast and convenient for extracting pore networks. However, when simulating a set of data, I found that the extracted pore network's `throat.direct_length = 0,` which is what you referred to pore and throat sizes were wonky. Since my pore network is randomly generated, I tried adjusting the relevant parameters. When I set voxel_size to 1, throat.direct_length is not 0, and I can calculate throat.hydraulic_conductance. Can you provide some guidance on further modifications?Here is my code:
`resolution = 1e-6
net = ps.networks.snow2(phases = im, voxel_size=resolution, r_max = 2, sigma = 0.2, boundary_width = 1)

pn = op.io.network_from_porespy(net.network)
pn.add_model(propname='pore.cluster_number',
             model=op.models.network.cluster_number)
pn.add_model(propname='pore.cluster_size',
             model=op.models.network.cluster_size)
pn["pore.diameter"] = pn['pore.inscribed_diameter']

Ps = pn['pore.cluster_size'] < 5552
op.topotools.trim(network=pn, pores=Ps)

water = op.phase.Water(network=pn)
R = pn['throat.inscribed_diameter']/2
L = pn['throat.direct_length']
mu = water['throat.viscosity'].max()
water['throat.hydraulic_conductance'] = np.pi*R**4/(8*mu*L)`

**Reply by LilHang** [reply]
Hi, `direct_length` you refer is the distance between the two centroids of adjacent pores, maybe some of your pores are too small and overlap a lot so the calculated `direct_length`is close to zero. I saw you use hagen-poiseuille equation directly to calculate conductance, that means you assume pressure drop only occurs in throats, and throats are cylinder. There are some other better options to calculate condunctance (see `predict_permeability `and `hydraulice_size_factors` tutorials on the openPNM website).

**Comment by LilHang** [comment]
UPDATE:
I use `openpnm.models.geometry.conduit_lengths`function and find it is some negative `L1` and `L2` values which are returned by this function that caused some negative size factors, and eventually result in some negative conductance. As professor Gostick explains, this may due to  pore and throat sizes are wonky, and overlap a lot. One can see much cleary by referring to the formula in this function source code , it is where the negative sign comes in. So i switch to `hybrid_cones_and_cylinders` which is a mix of `cones_and_cylinders `and `intersecting_cones` to calculate size factors, and there is no more negative conductance,the final results are much better than `cones_and_cylinders`.

**Reply by utopiankk** [reply]
Hello, may I ask how to use `hybrid_cones_and_cylinders` instead of `cones_and_cylinders`? I encountered the error: KeyError: 'throat.coords'.

**Reply by LilHang** [reply]
Sorry for the late response. I use `SNOW` to extract the pore network, so i assign the `throat.coords` property by `pn["throat.coords"] = pn["throat.global_peak"]`

**Comment by likeduck** [comment]
As far as I understand, SNOW is a watershed-based algorithm that is sensitive to the structure of the image, so complex structures will lead to a large number of pores in a small volume range, which may be the reason for throat=0. Therefore, it will be better to use SNOW to extract some processed images.
